storage/reader.go

```diff
@@ -80,6 +83,110 @@ func (o *ObjectHandle) NewReader(ctx context.Context) (*Reader, error) {
 	return o.NewRangeReader(ctx, 0, -1)
 }
 
+// newRangeReaderWithGRPC creates a new Reader with the given range that uses
+// gRPC to read Object content.
+//
+// This is an experimental API and not intended for public use.
+func (o *ObjectHandle) newRangeReaderWithGRPC(ctx context.Context, offset, length int64) (r *Reader, err error) {
+	ctx = trace.StartSpan(ctx, "cloud.google.com/go/storage.Object.newRangeReaderWithGRPC")
+	defer func() { trace.EndSpan(ctx, err) }()
+
+	if o.c.gc == nil {
+		err = fmt.Errorf("handle doesn't have a gRPC client initialized")
+		return
+	}
+	if err = o.validate(); err != nil {
+		return
+	}
+
+	// A negative length means "read to the end of the object", but the
+	// read_limit field it corresponds to uses zero to mean the same thing. Thus
+	// we coerce the length to 0 to read to the end of the object.
+	if length < 0 {
+		length = 0
+	}
+
+	req := &storagepb.ReadObjectRequest{
+		Bucket: o.bucket,
+		Object: o.object,
+	}
+
+	// Define a function that initiates a Read with offset and length, assuming we
+	// have already read seen bytes.
+	reopen := func(seen int64) (storagepb.Storage_ReadObjectClient, context.CancelFunc, error) {
+		// If the context has already expired, return immediately without making a
+		// call.
+		if err := ctx.Err(); err != nil {
+			return nil, nil, err
+		}
+
+		cc, cancel := context.WithCancel(ctx)
+
+		start := offset + seen
```

Comment:
Does the golang library support reading from a negative offset?  It is possible to ask GCS to read "starting at byte -200" which really means "read the lasts 200 bytes".  If that is supported, reading from "offset + 200" is not the same as reading from "-0", I would leave it up to you to figure this out.

Reply:
> Does the golang library support reading from a negative offset?

Yes it should. I had to make some code changes to support this, so thank you. Initializing the `Reader.size` and `Reader.remain` to the `abs(offset)` so that the reader will stop once it has seen that `size`. Haven't pushed the changes yet b.c I'm having trouble with CRCs now.

> "starting at byte -200"...reading from "offset + 200" is not the same as reading from "-0"

I don't think this will be an issue because of the aforementioned logic. We should not attempt to reopen if the `Reader.size` and `Reader.seen` are equal. In your example, the `Reader.size` would be `200` and the `Reader.seen` should never surpass `200`. I could put a check in there, or a note warning of this, but I _think_ the logic is sound.

Reply:
Ok, I've pushed the changes with a test.